source/module_elecstate/elecstate_energy.cpp

```diff
@@ -274,7 +277,7 @@ void ElecState::cal_energies(const int type)
         this->f_en.esol_cav = get_solvent_model_Acav();
     }
 #ifdef __LCAO
-    if (GlobalV::sc_mag_switch)
+    if (SpinConstrain<std::complex<double>, psi::DEVICE_CPU>::get_sc_mag_switch())
```

Comment:
Could you explain why SpinConstrain is a template, and why it relates to FPTYPE and DEVICE?

Reply:
SpinConstrain is a template because it uses hsolver:
```
 hsolver::HSolver<FPTYPE, Device>* phsol = nullptr;
```